Support mix of int32 and int64 offsets/indices for EmbeddingBag and its variants

zyan0 opened the pull request:
Summary: Currently EmbeddingBag and it variants support either int32 or int64 indices/offsets. We have use cases where there are mix of int32 and int64 indices which are not supported yet. To avoid introducing too many branches we could simply cast offsets type to indices type when they are not the same.

Test Plan: unit tests

Differential Revision: D26820202

jbschlosser commented:
Hey @zyan0 ! Is there an associated issue for this PR?

jbschlosser commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -433,20 +433,25 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> _embedding_bag_cpu_impl(
     bool requires_grad) {
+  Tensor offsets_ = offsets;
+  if (offsets_.scalar_type() != indices.scalar_type()) {
+    offsets_ = offsets_.toType(indices.scalar_type());
```
FYI it's probably better to use type promotion for these conversions (check out `promoteTypes()`), as that avoids potential issues going from e.g. int32 to int8.

zyan0 commented:
> Hey @zyan0 ! Is there an associated issue for this PR?

Thanks for reviewing! I don't think there is a relevant issue. Should I create one and associate with this PR?

Commit:
Support mix of int32 and int64 offsets/indices for EmbeddingBag and its variants (#53655)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/53655

Currently EmbeddingBag and it variants support either int32 or int64 indices/offsets. We have use cases where there are mix of int32 and int64 indices which are not supported yet. To avoid introducing too many branches we could simply cast offsets type to indices type when they are not the same.

Test Plan: unit tests

Reviewed By: qizzzh

Differential Revision: D26820202

fbshipit-source-id: 04e12cc1a36e0ce77cdf0bb836c6a76ae3b1ad57
